src/Controller/Backend/Authentication.php

```diff
@@ -59,10 +60,19 @@ public function getLogin(Request $request, $resetCookies = false)
         }
 
         $response = $this->render('@bolt/login/login.twig', ['randomquote' => true]);
-        $response->setVary('Cookies', false)->setMaxAge(0)->setPrivate();
-
-        if ($resetCookies) {
-            $response->headers->clearCookie($this->app['token.authentication.name']);
+        if ($response instanceof TemplateView) {
+            $this->app->after(function (Request $request, Response $response) use ($resetCookies) {
+                $response->setVary('Cookies', false)->setMaxAge(0)->setPrivate();
+                if ($resetCookies) {
+                    $response->headers->clearCookie($this->app['token.authentication.name']);
+                }
+            });
+        } else {
+            /** @deprecated Deprecated since 3.3, to be removed in 4.0. */
+            $response->setVary('Cookies', false)->setMaxAge(0)->setPrivate();
+            if ($resetCookies) {
+                $response->headers->clearCookie($this->app['token.authentication.name']);
+            }
```

Comment:
This should be changed to a route middleware too. You can just check if the `$response->isRedirect()` before modifying headers.